test/integration/src/test/java/io/pravega/test/integration/ReadWriteTest.java

```diff
@@ -0,0 +1,262 @@
+import io.pravega.segmentstore.server.host.handler.PravegaConnectionListener;
+import io.pravega.segmentstore.server.store.ServiceBuilder;
+import io.pravega.segmentstore.server.store.ServiceBuilderConfig;
+import io.pravega.test.common.TestUtils;
+import io.pravega.test.common.TestingServerStarter;
+import io.pravega.test.integration.demo.ControllerWrapper;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.test.TestingServer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.junit.Assert.assertEquals;
+
+@Slf4j
+public class ReadWriteTest {
+
+    private static final String STREAM_NAME = "testMultiReaderWriterStream" + new Random().nextInt(Integer.MAX_VALUE);
+    private static final int NUM_WRITERS = 5;
+    private static final int NUM_READERS = 5;
+    private static final long NUM_EVENTS = 500;
+    private AtomicBoolean stopReadFlag;
+    private AtomicLong eventData;
+    private AtomicLong eventReadCount;
+    private ConcurrentLinkedQueue<Long> eventsReadFromPravega;
+    private TestingServer zkTestServer = null;
+    private PravegaConnectionListener server = null;
+    private ControllerWrapper controllerWrapper = null;
+    private Controller controller = null;
+    private int i;
+
+    @Before
+    public void setup() throws Exception {
+
+        final int controllerPort = TestUtils.getAvailableListenPort();
+        final String serviceHost = "localhost";
+        final int servicePort = TestUtils.getAvailableListenPort();
+        final int containerCount = 4;
+
+        try {
+            // 1. Start ZK
+            this.zkTestServer = new TestingServerStarter().start();
+
+            // 2. Start Pravega service.
+            ServiceBuilder serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
+            serviceBuilder.initialize();
+            StreamSegmentStore store = serviceBuilder.createStreamSegmentService();
+
+            this.server = new PravegaConnectionListener(false, servicePort, store);
+            this.server.startListening();
+
+            // 3. Start  controller service
+            this.controllerWrapper = new ControllerWrapper(zkTestServer.getConnectString(), false,
+                    controllerPort, serviceHost, servicePort, containerCount);
+            this.controllerWrapper.awaitRunning();
+            this.controller = controllerWrapper.getController();
+        } catch (Exception e) {
+            log.error("Error during setup", e);
+            throw e;
```

Comment:
Assert fail eats up the real exception and stack trace, the `log` statement will not appear in the build logs, so we have no debug information if test fails here. Hence we should rethrow, or better, not catch the exception because we are not doing anything useful with it.